File: wasm/src/TecEnum.hs
module TecEnum where

import Data.Aeson
  ( FromJSON,
    ToJSON (toEncoding),
    defaultOptions,
    genericToEncoding,
  )
import GHC.Generics (Generic)


data TecValue = TecValue String deriving (Show, Generic)

instance ToJSON TecValue where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON TecValue

data TecEnum = TecEnum {tecEnumName :: String, tecEnumValues :: [TecValue]}
  deriving (Show, Generic)

instance ToJSON TecEnum where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON TecEnum

data TecEnumAST = TecEnumAST {tecEnums :: [TecEnum]} deriving (Show, Generic)

instance ToJSON TecEnumAST where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON TecEnumAST

File: web/libs/codec/src/TecEnum.ts
import { Schema as S } from "effect";
export const TecEnumValue = S.String.pipe(S.brand("TecEnumValue"));
export type TecEnumValue = typeof TecEnumValue.Type;

export const TecEnumName = S.String.pipe(S.brand("TecEnumName"));
export type TecEnumName = typeof TecEnumName.Type;

export const TecEnum = S.Struct({
  tecEnumName: TecEnumName,
  tecEnumValues: S.Array(TecEnumValue),
});
export type TecEnum = typeof TecEnum.Type;
export const TecEnumAST = S.Struct({ tecEnums: S.Array(TecEnum) });
export type TecEnumAST = typeof TecEnumAST.Type;
